axiom/upgrade.py

```diff
@@ -253,6 +254,81 @@ def upgradeAllTheWay(o):
     return o
 
 
+
+def _hasExplicitOid(store, table):
+    """
+    Does the given table have an explicit oid column?
+    """
+    return any(info[1] == 'oid' for info
+               in store.querySchemaSQL(
+                   'PRAGMA *DATABASE*.table_info({})'.format(table)))
+
+
+
+def _upgradeTableOid(store, table, createTable, postCreate=lambda: None):
+    """
+    Upgrade a table to have an explicit oid.
+
+    Must be called in a transaction to avoid corrupting the database.
+    """
+    if _hasExplicitOid(store, table):
+        return
+    store.executeSchemaSQL(
+        'ALTER TABLE *DATABASE*.{0} RENAME TO {0}_temp'.format(table))
+    createTable()
+    store.executeSchemaSQL(
+        'INSERT INTO *DATABASE*.{0} '
+        'SELECT oid, * FROM *DATABASE*.{0}_temp'.format(table))
+    store.executeSchemaSQL('DROP TABLE *DATABASE*.{0}_temp'.format(table))
+    postCreate()
+
+
+
+def upgradeSystemOid(store):
+    """
+    Upgrade the system tables to use explicit oid columns.
+    """
+    store.transact(
+        _upgradeTableOid, store, 'axiom_types',
+        lambda: store.executeSchemaSQL(CREATE_TYPES))
+    store.transact(
+        _upgradeTableOid, store, 'axiom_objects',
+        lambda: store.executeSchemaSQL(CREATE_OBJECTS),
+        lambda: store.executeSchemaSQL(CREATE_OBJECTS_IDX))
+
+
+
+def upgradeExplicitOid(store):
+    """
+    Upgrade a store to use explicit oid columns.
```

Comment:
This should probably say something about the fact that it makes it safe to VACUUM.